File: client/src/app/detector/de_wm.py
import os


def detect_de_wm() -> str | None:
    checks = [
        ("HYPRLAND_INSTANCE_SIGNATURE", "hyprland"),
        ("SWAYSOCK", "sway"),
        ("GNOME_DESKTOP_SESSION_ID", "gnome"),
        ("KDE_FULL_SESSION", "kde"),
    ]

    for env_var, name in checks:
        if os.getenv(env_var):
            return name

    xdg = os.getenv("XDG_CURRENT_DESKTOP", "").lower()
    if xdg:
        return xdg

    return None

File: client/internal/detector/dewm.go
package detector

import (
	"os"
	"strings"
)

var envChecks = []struct {
	env  string
	name string
}{
	{"HYPRLAND_INSTANCE_SIGNATURE", "hyprland"},
	{"SWAYSOCK", "sway"},
	{"GNOME_DESKTOP_SESSION_ID", "gnome"},
	{"KDE_FULL_SESSION", "kde"},
}

func detectDEWM() string {
	for _, c := range envChecks {
		if os.Getenv(c.env) != "" {
			return c.name
		}
	}

	if xdg := strings.ToLower(os.Getenv("XDG_CURRENT_DESKTOP")); xdg != "" {
		return xdg
	}

	return ""
}
